sdk/log/logger_test.go

```diff
@@ -348,3 +357,94 @@ func TestLoggerEnabled(t *testing.T) {
 		})
 	}
 }
+
+func TestLoggerSelfObservability(t *testing.T) {
+	testCases := []struct {
+		name                     string
+		selfObservabilityEnabled bool
+		ctx                      context.Context
+		records                  []log.Record
+		wantLogRecordCount       int64
+	}{
+		{
+			name:                     "SelfObservabilityDisabledZeroLogs",
+			selfObservabilityEnabled: false,
+			ctx:                      context.Background(),
+			records:                  nil,
+			wantLogRecordCount:       0,
+		},
+		{
+			name:                     "SelfObservabilityDisabledNonZeroLogs",
+			selfObservabilityEnabled: false,
+			ctx:                      context.Background(),
+			records:                  []log.Record{{}, {}},
+			wantLogRecordCount:       0,
+		},
+		{
+			name:                     "ZeroLogsCreated",
+			selfObservabilityEnabled: true,
+			ctx:                      context.Background(),
+			records:                  []log.Record{},
+			wantLogRecordCount:       0,
+		},
+		{
+			name:                     "NilContext",
+			selfObservabilityEnabled: true,
+			ctx:                      nil,
+			records:                  []log.Record{{}, {}, {}, {}},
+			wantLogRecordCount:       4,
+		},
```

Comment:
Can we remove this use case? 
Especially that handling `nil` is not something that we need to keep.
From https://pkg.go.dev/context#pkg-overview:

> Do not pass a nil [Context](https://pkg.go.dev/context#Context), even if a function permits it.